### PR TITLE
Remove blocking I/O from ReactPHP implementation

### PR DESCRIPTION
**Pull request by lcobucci**
The existing ReactPHP implementation doesn't leverage the async capabilities, creating wrong results when benchmarking things.

This optimises the ReactPHP app, allowing things to be a bit more comparable.

### CODE REVIEW

**Comment by lcobucci** [comment]
@WyriHaximus please feel free to correct things even further :+1:

File: frameworks/PHP/reactphp/server.php
Comment (WyriHaximus) [inline frameworks/PHP/reactphp/server.php:21]: No need to remove all the listeners, that will be done the moment the socket is closed: 
```suggestion
```

File: frameworks/PHP/reactphp/app.php
Comment (WyriHaximus) [inline frameworks/PHP/reactphp/app.php:25]: Seems like a waste creating these every request, maybe a `global` or a singleton *shivers* is in order here?

File: frameworks/PHP/reactphp/reactphp.dockerfile
Comment (WyriHaximus) [inline frameworks/PHP/reactphp/reactphp.dockerfile:31]: Shameless plug maybe, but you have tried my ReactPHP focussed docker images? Also comes with JIT turned on out of the box for a nice boost. Here is a multi stage example so you don't have your dev tools in your final image: https://github.com/wyrihaximusnet/docker-php#example-usage

Just a suggestion FYI.

**Review by WyriHaximus — CHANGES_REQUESTED** [review]

File: frameworks/PHP/reactphp/app.php
Comment (lcobucci) [reply frameworks/PHP/reactphp/app.php:25]: This isn't being created on each request, though.

The function scope and returned closure behaves as a singleton.

File: frameworks/PHP/reactphp/app.php
Comment (lcobucci) [reply frameworks/PHP/reactphp/app.php:17]: Not really, the lazy connection is created only once and reused everywhere

File: frameworks/PHP/reactphp/reactphp.dockerfile
Comment (lcobucci) [reply frameworks/PHP/reactphp/reactphp.dockerfile:31]: I didn't want to change everything too drastically. If you think it would have huge impact on the benchmark, then maybe it'd make sense. However, I think we could/should defer that to another PR.

Comment (WyriHaximus) [reply frameworks/PHP/reactphp/reactphp.dockerfile:31]: Yeah I get that, hence putting it in as suggestion :)

File: frameworks/PHP/reactphp/app.php
Comment (WyriHaximus) [reply frameworks/PHP/reactphp/app.php:17]: Pretty sure reading the code all the way up https://github.com/friends-of-reactphp/mysql/blob/0.7.x/src/Io/LazyConnection.php there is no way it gets reused. Unless I'm missing something.

Comment (lcobucci) [reply frameworks/PHP/reactphp/app.php:17]: I'll run tests to verify it but I'm pretty sure the reuse of `$connection` here combined with an 500ms idle limit makes `LazyConnection` to not open new internal connections in `LazyConnection#connecting()`

File: frameworks/PHP/reactphp/app.php
Comment (WyriHaximus) [reply frameworks/PHP/reactphp/app.php:25]: :facepalm: Just noticed it's returning a closure, you're right about here and `$connection`. Ignore this you're right :+1: . Going to feel stupid for a couple of minutes :joy: .

Comment (lcobucci) [reply frameworks/PHP/reactphp/app.php:25]: HAHAHAHA I imagined that the closure was indeed being missed from your review 😂

There's no need for feeling stupid, I can't think of anyone who never overlooked things ❤️

File: frameworks/PHP/reactphp/app.php
Comment (lcobucci) [reply frameworks/PHP/reactphp/app.php:17]: Just confirmed that we're indeed keeping the connection open (due to the closure thingy and the idle limit).
Sent several requests to the `/db` endpoint and `show processlist` in MySQL displayed the same PID for all requests.

File: frameworks/PHP/reactphp/app.php
Comment (joanhey) [reply frameworks/PHP/reactphp/app.php:25]: For good results, here we need to only execute the prepared statement.

Comment (WyriHaximus) [reply frameworks/PHP/reactphp/app.php:25]: > For good results, here we need to only execute the prepared statement.

Even better. 0.7 will come with a connection pool: https://github.com/friends-of-reactphp/mysql/issues/175

**Comment by joanhey** [comment]
Please update to PHP 8.3
- #8579

**Comment by lcobucci** [comment]
> Please update to PHP 8.3
> - #8579

Sure, will do around next week

**Comment by lcobucci** [comment]
@joanhey a6657b758d397c73ccb60201403de52af0d9daca handles the PHP 8.3 upgrade :+1:

**Comment by joanhey** [comment]
@lcobucci 
Could you check ReactPHP, as it's failing for a long time.
https://tfb-status.techempower.com/unzip/results.2025-01-04-12-20-28-252.zip/reactphp

**Comment by WyriHaximus** [comment]
@joanhey Just checked those logs but don't see any obvious failures there. I will have a look at the code and docker file locally and see if I can spot what's wrong there.

**Comment by WyriHaximus** [comment]
@joanhey Do you remember when it started breaking? Because it's a MySQL related issue, so one of these 3 broke it; https://github.com/TechEmpower/FrameworkBenchmarks/pull/8715, https://github.com/TechEmpower/FrameworkBenchmarks/pull/9009, or https://github.com/TechEmpower/FrameworkBenchmarks/pull/9183.

**Comment by joanhey** [comment]
I can check when started to fail.
But it's more important, if anything fail than show the problem to the devs.

**Comment by joanhey** [comment]
We update Mysql from 5.7 to 8.0 and now to 9.0,
Anybody can check in local this.
But for me, the worst think is than don't show where is the problem the framework.

**Comment by joanhey** [comment]
If it fail for the Mysql, it's OK.
But why fail the tests without DB ?

**Comment by joanhey** [comment]
Like I said, problems always exist, but the more important  think is that the language or framework, show you where is the problem.

**Comment by joanhey** [comment]
Here started to fail:
May 7, 2024

image

I'll investigate what changed in the bench, but please also search what changed in ReactPHP !!

**Comment by joanhey** [comment]
Changes in the toolset :
https://github.com/TechEmpower/FrameworkBenchmarks/commits/master/toolset

Changes in the benchmark code for ReactPHP:
https://github.com/TechEmpower/FrameworkBenchmarks/commits/master/frameworks/PHP/reactphp

**Comment by joanhey** [comment]
The last run that work ReactPHP:
Here we show the last commits 
https://github.com/TechEmpower/FrameworkBenchmarks/commits/39e29cb04e7d78074ee363aea1e8e2e3d738cbc9/

The next run that failed:
https://github.com/TechEmpower/FrameworkBenchmarks/commits/c3e4826cf93907ed0175c7d8fc09488ab1fb4167/

I don't see anything bad in the commits.
Please check it !!

**Comment by joanhey** [comment]
A good benchmark it isn't only for performance checks.
I use it to fix problems too.

Check: #9408

**Comment by WyriHaximus** [comment]
> I can check when started to fail. But it's more important, if anything fail than show the problem to the devs.

Yes, which is why I just filed: https://github.com/TechEmpower/FrameworkBenchmarks/pull/9527

**Comment by WyriHaximus** [comment]
> Changes in the toolset : https://github.com/TechEmpower/FrameworkBenchmarks/commits/master/toolset
> 
> Changes in the benchmark code for ReactPHP: https://github.com/TechEmpower/FrameworkBenchmarks/commits/master/frameworks/PHP/reactphp

Yeah checked those, had to put in a crude version of https://github.com/TechEmpower/FrameworkBenchmarks/pull/9527 to find the issue.
 
> PD: if you check the commits, you will see than I try my best, but it is not enough sometimes

It's a lot, and we're thankful for that.

> I'll investigate what changed in the bench, but please also search what changed in ReactPHP !!

Nothing actually, it was the MySQL upgrades to a version with the auth method ReactPHP currently supports enabled. (We need to add support for the never ones 😅 .)

**Comment by WyriHaximus** [comment]
> If it fail for the Mysql, it's OK. But why fail the tests without DB ?

Because the connectivity test is done against the `/db` endpoint instead of `/plaintext`, not sure how to control that so filed  https://github.com/TechEmpower/FrameworkBenchmarks/pull/9528 as a step in the right direction. Because having this as out coming makes it a lot clearer where the problem might be:
image

**Comment by joanhey** [comment]
You can test it locally
`./tfb --mode verify --test reactphp`

And if you want to test it directly calling the urls:
`./tfb --mode debug --test reactphp`

**Comment by joanhey** [comment]
After add #9527

image

**Comment by joanhey** [comment]
#9527 and #9528 never will be merged, as don't pass the tests.

**Comment by joanhey** [comment]
Plaintext and json work without problems.
If you want, we can disable temporally the db tests till fixed.

**Comment by WyriHaximus** [comment]
> Plaintext and json work without problems. If you want, we can disable temporally the db tests till fixed.

Lets start there, will file a PR for that 👍

**Comment by WyriHaximus** [comment]
Given #8303 I've opened #9533

**Comment by WyriHaximus** [comment]
> #9527 and #9528 never will be merged, as don't pass the tests. You need to test it locally till fix the problem.

FWIW: https://github.com/TechEmpower/FrameworkBenchmarks/pull/8715, https://github.com/TechEmpower/FrameworkBenchmarks/pull/9009, and  https://github.com/TechEmpower/FrameworkBenchmarks/pull/9183 all got merged with failing tests. 

> Plaintext and json work without problems. If you want, we can disable temporally the db tests till fixed.

Yup just filed https://github.com/TechEmpower/FrameworkBenchmarks/pull/9534 for that, will rebase the others once that is passing and merged

**Comment by joanhey** [comment]
I created #9529 yesterday, now it's duplicated.

About the Databases update, normally fail, because always exist frameworks that fail.
With the Databases we test than run with the frameworks than normally work.

**Comment by WyriHaximus** [comment]
> I created #9529 yesterday, now it's duplicated.

Missed that as it wasn't mentioned in the PR title, sorry.

> About the Databases update, normally fail, because always exist frameworks that fail. Even frameworks than don't use the database. With the Databases we test than run with the frameworks than normally work.

As a maintainer I get it, especially something this big I get it. And I get that #9528 will never be merged, that thing was a discussion starter. However, #9527 is perfectly fine to merge as it doesn't impact the benchmark negatively and will actually positively aid in finding issues with it.

> PD: the frameworks failing in the last run https://tfb-status.techempower.com/results/924216cc-3d66-43a2-ad8b-4d423e968e3e

Ow sweet thanks, it's kinda good to see we're not the only ones 😅 .

**Comment by joanhey** [comment]
#9257 only show the error, but don't fix the problem.

The tests will fail, and no framework is merged if don't pass the tests in the PR.

**Comment by joanhey** [comment]
I don't mentioned in the title, because the important thing is the update to PHP8.4.
But I comment it that it's necessary disable the db tests and fix the broken run.

**Comment by WyriHaximus** [comment]
> #9257 only show the error, but don't fix the problem.
> 
> The tests will fail, and no framework is merged if don't pass the tests in the PR.

#9538 will not fix the database issue, but at least respond with an error

**Comment by WyriHaximus** [comment]
> I don't mentioned in the title, because the important thing is the update to PHP8.4. But I comment it that it's necessary to disable the db tests to fix the broken run.

Fair, I was just expecting it in it's own dedicated PR. But that's up to you as maintainer how you prefer to handle that :)

**Pull request merged**
